### PR TITLE
Setup Shipment initially

### PR DESCRIPTION
**Pull request by nk-coding**
https://frontend.gropius.duckdns.org/components/b004efba-b36f-40c2-9b49-4e4eed850335/issues/2cd30ee1-f88e-4136-8f62-1dfa7c9eec71

- currently still missing correct integration with `payment/payment/payment-enabled` as the exact schema of the context passed ~~is unknown~~ not fully specified yet
- partially untested, depends on order and discount

DoD:
- [x] Requirements of the issue are met
- [x] Code has been reviewed
- [x] Code is documented
- [x] GraphQL related artefacts are documented

### CODE REVIEW

File: docker-compose-dev.yaml [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    extends:
+      file: docker-compose-base.yaml
+      service: shipment-dapr
+    volumes:
+      - "./.dapr/dapr-config-minimal.yaml:/config.yaml"
+      - "./.dapr/components:/components"
+  debug-shipment-provider:
+    image: "mendhak/http-https-echo:31"
+    environment:
+      - ECHO_BACK_TO_CLIENT=false
```
Comment (delvh) [inline docker-compose-dev.yaml:26]: Do we have a debug container in any other service?

File: src/main/resources/db/migration/V1_1__create_relations.sql [hunk 2]
```diff
@@ -0,0 +1,54 @@
+CREATE EXTENSION IF NOT EXISTS "uuid-ossp";
+
+CREATE TABLE AddressEntity (
```
Comment (delvh) [inline src/main/resources/db/migration/V1_1__create_relations.sql:3]: If we duplicate the entire address service into this service, wouldn't it make sense to consolidate them?

File: src/main/kotlin/org/misarch/shipment/graphql/GraphQLConfiguration.kt [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+                return when (type.classifier) {
+                    OffsetDateTime::class -> DateTimeScalar.INSTANCE
+                    UUID::class -> UUIDScalar.INSTANCE
+                    else -> null
+                }
+            }
+
+            override fun didBuildSchema(builder: GraphQLSchema.Builder): GraphQLSchema.Builder {
+                return super.didBuildSchema(builder).clearDirectives()
+            }
```
Comment (delvh) [inline src/main/kotlin/org/misarch/shipment/graphql/GraphQLConfiguration.kt:31]: Did I miss something in the previous services, or was this block missing there?

File: src/main/kotlin/org/misarch/shipment/graphql/model/connection/ShipmentMethodConnection.kt [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+@GraphQLDescription("Shipment method filter")
+class ShipmentMethodFilter(
+    @GraphQLDescription("Filter shipment methods by their archived status")
+    val isArchived: Boolean?
+) : BaseFilter {
+
+    override fun toExpression(): BooleanExpression? {
+        return if (isArchived != null) {
+            ShipmentMethodEntity.ENTITY.archivedAt.let {
+                if (isArchived) it.isNotNull else it.isNull
```
Comment (delvh) [inline src/main/kotlin/org/misarch/shipment/graphql/model/connection/ShipmentMethodConnection.kt:81]: I am confused. Didn't you already ensure that `is Archived != null`?

File: src/main/kotlin/org/misarch/shipment/graphql/model/Shipment.kt [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+    @GraphQLDescription("The address this shipment is sent to.")
+    fun shipmentAddress(
+        dfe: DataFetchingEnvironment
+    ): CompletableFuture<Address> {
+        return dfe.getDataLoader<UUID, Address>(AddressDataLoader::class.simpleName!!)
+            .load(shipmentAddressId, dfe)
+    }
+
+    @GraphQLDescription("The order this shipment is caused by.")
```
Comment (delvh) [inline src/main/kotlin/org/misarch/shipment/graphql/model/Shipment.kt]:
```suggestion
    @GraphQLDescription("The order creating this shipment.")
```

File: src/main/kotlin/org/misarch/shipment/graphql/model/Shipment.kt [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+    fun order(): Order? {
+        return orderID?.let { Order(it) }
+    }
+
+    @GraphQLDescription("The return this shipment is caused by.")
```
Comment (delvh) [inline src/main/kotlin/org/misarch/shipment/graphql/model/Shipment.kt]:
```suggestion
    @GraphQLDescription("The product return causing this shipment.")
```

File: src/main/kotlin/org/misarch/shipment/graphql/model/ShipmentMethod.kt [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+    val archivedAt: OffsetDateTime?, private val ref: String
+) : Node(id) {
+
+    @GraphQLDescription("If true, this shipment method is archived and can no longer be used.")
+    val isArchived: Boolean
+        get() = archivedAt != null
+
+
+    @GraphQLDescription("The reference of the shipment method used by the external shipment provider.")
+    fun ref(dfe: DataFetchingEnvironment): String {
```
Comment (delvh) [inline src/main/kotlin/org/misarch/shipment/graphql/model/ShipmentMethod.kt]: Why shorten it in GraphQL?

Comment (delvh) [reply src/main/kotlin/org/misarch/shipment/graphql/model/ShipmentMethod.kt]: I.e. use `externalReference` instead

File: src/main/resources/db/migration/V1_1__create_relations.sql (hunk 2 above)
Comment (delvh) [reply src/main/resources/db/migration/V1_1__create_relations.sql:3]: Discussed privately

File: docker-compose-dev.yaml (hunk 1 above)
Comment (nk-coding) [reply docker-compose-dev.yaml:26]: no
but the thing is, shipment needs an external shipment provider where shipments can be registered
without this, shipment stuff could not be tested
however, if you want, I can remote it

File: src/main/kotlin/org/misarch/shipment/graphql/GraphQLConfiguration.kt (hunk 3 above)
Comment (nk-coding) [reply src/main/kotlin/org/misarch/shipment/graphql/GraphQLConfiguration.kt:31]: e.g. https://github.com/MiSArch/address/blob/main/src/main/kotlin/org/misarch/address/graphql/GraphQLConfiguration.kt#L29

Comment (nk-coding) [reply src/main/kotlin/org/misarch/shipment/graphql/GraphQLConfiguration.kt:31]: fyi, this code is to get rid of not needed federation directives

File: src/main/kotlin/org/misarch/shipment/graphql/model/connection/ShipmentMethodConnection.kt (hunk 4 above)
Comment (nk-coding) [reply src/main/kotlin/org/misarch/shipment/graphql/model/connection/ShipmentMethodConnection.kt:81]: yes I did
but it is still a boolean which can be true or false
if null is provided, no filter is applied
if true is provided, you only get archived ones
if false is provided, you only get non-archived ones

File: src/main/kotlin/org/misarch/shipment/graphql/model/ShipmentMethod.kt (hunk 7 above)
Comment (nk-coding) [reply src/main/kotlin/org/misarch/shipment/graphql/model/ShipmentMethod.kt]: changed with 86acba666881a13db997f7a380f5bee66033ceed

File: docker-compose-dev.yaml (hunk 1 above)
Comment (delvh) [reply docker-compose-dev.yaml:26]: No, makes sense

File: src/main/kotlin/org/misarch/shipment/graphql/GraphQLConfiguration.kt (hunk 3 above)
Comment (delvh) [reply src/main/kotlin/org/misarch/shipment/graphql/GraphQLConfiguration.kt:31]: > e.g. https://github.com/MiSArch/address/blob/main/src/main/kotlin/org/misarch/address/graphql/GraphQLConfiguration.kt#L29

Apparently I missed it across 7 PR reviews…

File: src/main/kotlin/org/misarch/shipment/graphql/model/connection/ShipmentMethodConnection.kt (hunk 4 above)
Comment (delvh) [reply src/main/kotlin/org/misarch/shipment/graphql/model/connection/ShipmentMethodConnection.kt:81]: Ah, true.
Makes sense.

**Review by delvh — APPROVED** [review]

**Pull request merged**